Add Send REST API endpoint

akaravo opened the pull request:
Add Send REST API endpoint

akaravo committed:
#746 added Send endpoint to create and publish campaigns from clients with client credentials

travlos committed:
Add Swagger to endpoint

cleftheris committed:
Merge branch 'develop' into feature/messaging/send

cleftheris commented on src/Indice.Features.Messages.Core/Models/Requests/SendRequest.cs:
The send request should be a marker class probably subclassing the CreateCampaignRequest rather than the CampaignRequestBase so that we do not need the duplicate code.

akaravo replied:
The issue with this is CreateCampaignRequest has a Published property (bool).
SendRequest : CreateCampaignRequest redefines it using new keyword to always return true.
But Swagger (via Swashbuckle) still picks up the base class's Published property and displays it in the API documentation for SendRequest.

We want to hide Published in SendRequest from Swagger UI because it will always be true. 
This is the one option for this to work , the other was with an ISchemaFilter.

If we don't want for the campaign to always be published when created from Send Api , then we can go with marker class.

cleftheris replied:
In this case we should make the published property virtual and in the subclass decorate it with a [JsonIgnore] attribute.

cleftheris commented on src/Indice.Features.Messages.Core/MessageEndpointOptions.cs:
```diff
@@ -36,6 +36,8 @@ public MessageManagementOptions() : base() { }
     /// <summary>The default scope name to be used for Messages API. Defaults to <see cref="MessagesApi.Scope"/>.</summary>
     public string RequiredScope { get; set; } = MessagesApi.Scope;
+    /// <summary>The default scope name to be used for Send API. Defaults to <see cref="MessagesApi.SendScope"/>.</summary>
+    public string SendRequiredScope { get; set; } = MessagesApi.SendScope;
```
I prefer a static constant for platform tools and platform scopes. It should not be configurable there is no benefit coming from this other than frustration on different installations. This is the way we went with cases, identity, messages, risk etc.

akaravo replied:
I will remove SendRequiredScope from MessageEndpointOptions.cs since it is not really required anymore, and it will be handled in CanSendCampaignHandler.cs

cleftheris commented on src/Indice.Features.Messages.Core/Constants.cs:
```diff
@@ -9,7 +9,9 @@ public static class MessagesApi
     /// <summary>Messages API scope.</summary>
-    public const string Scope = "messages";
+    public const string Scope = "messages";    
+    /// <summary>Messages API scope.</summary>
```
fix the xml comment

cleftheris commented on src/Indice.Features.Messages.AspNetCore/Authorization/BeCampaignSendRequirement.cs:
```diff
@@ -0,0 +1,50 @@
+            return Task.CompletedTask;
+        }
+        // Get user id/application id from the corresponding claims.
+        var allowed =   context.User!.HasScope(MessagesApi.SendScope);
```
This must be handled via both 
1. a client with the new scope `messages:send` (regardless of user priviledges such as role or IsAdmin).
2. a user that has the flag `IsAdmin` or the role `CampaignsManager`  and the client has the scope `messages`

cleftheris commented:
@travlos I added my request for changes comments above. Hope this helps

cleftheris commented on src/Indice.Features.Messages.AspNetCore/Endpoints/SendApi.cs:
```diff
@@ -0,0 +1,47 @@
+                                           .RequireAuthenticatedUser()
+                                           .RequireClaim(BasicClaimTypes.Scope, allowedScopes));
+
+        group.AddOpenApiSecurityRequirement("oauth2", allowedScopes).WithOpenApiSecurityRequirement("oauth2", allowedScopes);
```
Must split the allowed scopes that will go 
1. into the `OpenApi` (aka swagger `AddOpenApiSecurityRequirement` & `WithOpenApiSecurityRequirement`) security requirement, 
2. from the authorization requirement:`.RequireClaim(BasicClaimTypes.Scope, allowedScopes))`

This means that the OpenApi requirements should have all posible scopes in order to be available for selection. The same is not true for the security requirement. There the scopes must be validated via the custom Auth Requirement we wrote `BeCampaignSendRequirement`

cleftheris commented on src/Indice.Features.Messages.AspNetCore/Authorization/BeCampaignSendRequirement.cs:
```diff
@@ -0,0 +1,50 @@
+namespace Indice.Features.Messages.AspNetCore.Authorization;
+
+/// <summary>This authorization requirement specifies that an endpoint must be accessible only to Messaging tool managers.</summary>
+public class BeCampaignSendRequirement : IAuthorizationRequirement
```
rename to CanSendCampaignRequirement

cleftheris commented on src/Indice.Features.Messages.AspNetCore/Authorization/BeCampaignSendRequirement.cs:
```diff
@@ -0,0 +1,50 @@
+}
+
+/// <summary>Authorization handler corresponding to the <see cref="BeCampaignSendRequirement"/>.</summary>
+public class BeCampaignSendHandler : AuthorizationHandler<BeCampaignSendRequirement>
```
Rename to CanSendCampaignHandler

cleftheris committed:
Merge branch 'develop' into feature/messaging/send

akaravo committed:
#754
Refactor campaign sending authorization logic

Removed `BeCampaignSendRequirement` and added `CanSendCampaignRequirement` for improved permission checks. Updated authorization policies with `RequireSendManagement`. Adjusted scope management in `SendApi` and `MessageManagementOptions`. Refactored `SendRequest` to inherit from `CreateCampaignRequest`, enhancing code clarity and maintainability.

akaravo committed:
Merge branch 'feature/messaging/send' of https://github.com/indice-co/Indice.Platform into feature/messaging/send

travlos committed:
Refactor authorization and update dependencies

Upgraded NuGet packages:
- `Microsoft.Azure.Functions.Worker.Extensions.Storage.Queues` to `5.5.3`.
- `Microsoft.Extensions.Azure` to `1.13.0`.

Enhanced OpenAPI configuration:
- Added both `OAuth2AuthorizationCodeFlow` and `ClientCredentials` for the "send" API.

Refactored authorization logic:
- Introduced `HasMessagingScope` and unified role checks in `CanSendCampaignHandler`.
- Renamed `RequireSendManagement` to `RequireCanSendMessage` for clarity.
- Updated `CampaignsApi` and `SendApi` to use improved policies.

Added a new scope `messages:send` in `appsettings.json` for better access control.

Improved code readability and maintainability by removing redundant methods and constants.

cleftheris commented on src/Indice.Features.Messages.Core/Models/Requests/SendRequest.cs:
```diff
@@ -0,0 +1,45 @@
+    /// Copies all relevant properties and sets <see cref="CreateCampaignRequest.Published"/> to true.
+    /// </summary>
+    /// <returns>A new <see cref="CreateCampaignRequest"/> instance containing the mapped data.</returns>
+    public CreateCampaignRequest ToCreateCampaignRequest() {
```
SendRequest is a campaign create request so this is redundant code

travlos committed:
Casting is not required

cleftheris committed:
Merge branch 'develop' into feature/messaging/send

cleftheris reviewed (approved):
🚀

cleftheris commented:
closes #754
